internal/controller/apiextensions/claim/configurator.go

```diff
@@ -136,13 +121,7 @@ func (c *APIDryRunCompositeConfigurator) Configure(ctx context.Context, cm resou
 	ucp.SetClaimReference(proposed)
 
 	if !meta.WasCreated(cp) {
-		// The API server returns an available name derived from
-		// generateName when we perform a dry-run create. This name is
-		// likely (but not guaranteed) to be available when we create
-		// the composite resource. If the API server generates a name
-		// that is unavailable it will return a 500 ServerTimeout error.
-		cp.SetGenerateName(fmt.Sprintf("%s-", cm.GetName()))
-		return errors.Wrap(c.client.Create(ctx, cp, client.DryRunAll), errName)
+		cp.SetName(names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", cm.GetName())))
```

Comment:
does a dry-run request verify that the name is not taken? I.e. do we lose anything with this?

Reply:
If the answer is yes, i.e. dry-run verifies, then this change means that
- before we risked only conflict when an equally named ob
ject was created between claim update and XP creation.
- now we risk conflict when any of the names conflict: 23 bits in 5 characters => more than 98% risk for 10k objects.

<img width="411" alt="Bildschirmfoto 2023-10-11 um 15 45 26" src="https://github.com/crossplane/crossplane/assets/730123/6fe76500-71c3-48ee-b847-b8784ab91197">

Reply:
Ways out: check whether the name is taken, either with a client (roundtrip again, but 🤷) or checking the cache.

Reply:
> does a dry-run request verify that the name is not taken? I.e. do we lose anything with this?

The [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values) advises not to use that, i.e. we should not rely on the generated names.

Looking at the code, I would say that in the dry-run case, we are not going to get an error if the generated name already exists: 

* apiserver uses the same name generator to generate `metadata.name`: https://github.com/kubernetes/apiserver/blob/bf71493219c2bffe59c0b8c8b497ca0aa6e91a5a/pkg/registry/generic/registry/store.go#L408
* later on, [`Storage.Create`](https://github.com/kubernetes/apiserver/blob/bf71493219c2bffe59c0b8c8b497ca0aa6e91a5a/pkg/registry/generic/registry/store.go#L448) is called, passing also potential dry-run option.
* in case of dry-run, `DryRunnableStorage` handles it like this: https://github.com/kubernetes/apiserver/blob/d2d2fc882803682a397b74a9a0635fb6c1ffeff1/pkg/registry/generic/registry/dryrun.go#L40-L45
  * retrieve the value under the given key from the storage
  * merge the stored value with the submitted one

> does a dry-run request verify that the name is not taken? I.e. do we lose anything with this?

Given the above reasoning, dry-run request does not verify it, and the change proposed in this PR have the equal semantic.

Reply:
> Ways out: check whether the name is taken, either with a client (roundtrip again, but 🤷) or checking the cache.

I think this is not needed: when a composite cannot be created, because the name is taken, we requeue - in the next iteration a new name will be generated that is hopefully not taken.

Reply:
> https://github.com/kubernetes/apiserver/blob/d2d2fc882803682a397b74a9a0635fb6c1ffeff1/pkg/registry/generic/registry/dryrun.go#L40-L45

Means that it does check, no?

Reply:
> Means that it does check, no?

Looking at https://github.com/kubernetes/apiserver/blob/bf71493219c2bffe59c0b8c8b497ca0aa6e91a5a/pkg/registry/generic/registry/store.go#L450C3-L450C70, after calling `Create` of the dry-runnable storage, which then returns AlreadyExists, this `err = rest.CheckGeneratedNameError(ctx, e.CreateStrategy, err, obj)` call turns the AlreadyExists into [a conflict error with 1s retry](https://github.com/kubernetes/apiserver/blob/master/pkg/registry/rest/create.go#L168). The client will do this retry and cause another name being selected, until there is no conflict.

Hence, I think we cannot move forward here without checking the names in the cache to at least not regress.

Reply:
> I think this is not needed: when a composite cannot be created, because the name is taken, we requeue - in the next iteration a new name will be generated that is hopefully not taken.

Missed that comment.

If it succeeds to create the XR, but then naturally requeues, the informer might be lagging and your get request at the top will return NotFound, leading to another creation aka a leak.

Reply:
This is in the binder:
```
	// We refuse to 're-bind' a claim that is already bound to a different
	// composite resource.
	if existing != nil && !equal {
		return errors.New(errBindClaimConflict)
	}
```
Doesn't it mean that this rebinding on requeue is actually not working, not even in the old code?

Reply:
> Hence, I think we cannot move forward here without checking the names in the cache to at least not regress.

Oh, I missed that code on client retry when the name is already taken, thanks for the clarification :)

> If it succeeds to create the XR, but then naturally requeues, the informer might be lagging and your get request at the top will return NotFound, leading to another creation aka a leak.

Hm, what is the scenario? I thought about this:

1. Initially, there is no XR, just the claim
2. we generate composite name and the reference to it into the claim
3. the claim gets updated
4. composite gets created
5. if everything successful, we do not requeue - the reconcile will be triggered by claim/composite change

Are you saying that  if step 4 is successful, the next reconcile will not see it and `client.Get` could return NotFound?

Error scenarios:
* step 3 fails, then no XR gets created and we can requeue and retry
* step 4 failed,no XR gets created, and by the next reconcile  `existing` would be `nil`, so the new name will be generated and the new ref set in the claim

What this PR really needs is just to call `client.Create` in cases when composite does not exist. Right now we use `client.Apply` both for creation and updates.

Reply:
> Are you saying that if step 4 is successful, the next reconcile will not see it and client.Get could return NotFound?

yes. Reads are either backed by the cache (because controller-runtime does that automatically), or they are gets against the apiserver. But for the later, you have to add `resourceVersion=0` in order to get a quorum read from etcd. Otherwise, it is also backed by a cache (watch cache). So a 404 can happen, yes.

> if everything successful, we do not requeue

We do: we update both the claim and the xr. Both trigger a requeue (in the case of the xr of course without 404).

Reply:
> What this PR really needs is just to call client.Create

Yes, would very much prefer to be very explicit in what we are doing, not hiding it behind some apply wrapper.